### ISSUE TITLE
Code Actions shortcuts

### ISSUE THREAD

**Issue by ovidius72** [zed-industries/zed#14470]
### Check for existing issues

- [X] Completed

### Describe the feature

Please provide a way to assign shortcuts to individual code actions (organize import, sort import, ....)
Thanks

**Comment by notpeter** [comment]
Are there other code actions beyond 'organize imports' that you'd like actions for that cannot be assigned to shortcuts? Otherwise I think this is a duplicate of:
- https://github.com/zed-industries/zed/issues/10004

**Comment by ovidius72** [comment]
> Are there other code actions beyond 'organize imports' that you'd like actions for that cannot be assigned to shortcuts? (Or is 'organize imports' distinct 'sort imports' in some way I don't understand?) Otherwise I think this is a duplicate of:
> 
> * [Organize Imports Shortcut - CMD + Shift + O #10004](https://github.com/zed-industries/zed/issues/10004)

There are other code actions other than Organize Imports.

Here is a screenshot for typescript react:

<details><summary>screenshot</summary>

![Image](https://github.com/user-attachments/assets/6e9800b1-b050-4363-a8f0-cba5c12a8b1b)

</details>

I think it would be good to be able to assign each of them to single shortcuts.

> Or is 'organize imports' distinct 'sort imports' in some way I don't understand?

Organize imports, as far as I know, should run `sort imports`, `remove unused imports` and `add all missing imports` imports` in sequence.

**Comment by suryasarisa99** [comment]
In VSCode, when the Refactor  menu is opened, then typing automatically selects action based on input match. 
This is extremely useful in case more code actions there. add that to zed

**Comment by polymorphicengine** [comment]
bumping this, as it would be incredibly useful in my usecase. i'm developing a custom lsp for live coding music. the lsp provides custom code actions that should be quickly accessible via configurable shortcuts. 
thank you!

**Comment by notpeter** [comment]
> @suryasarisa99 In VSCode, when the Refactor menu is opened, then typing automatically selects action based on input match.

I have a POC implementation of this for all Zed context menus including code actions. I need to figure out some performance issues before opening a PR.

---

> @polymorphicengine: i'm developing a custom lsp for live coding music.  the lsp provides custom code actions that should be quickly accessible via configurable shortcuts.

Very cool.  So you would be looking for the ability to define custom keymap entries to trigger  code actions (note my understanding is code actions are cursor position specific). So something like:

```json
  {
    "context": "Editor && mode == full",
    "bindings": {
      "ctrl-alt-1": ["editor::CodeAction", "Move to a new file"],
      "ctrl-alt-+": ["editor::CodeAction", "Pitch Up"],
      "ctrl-alt--": ["editor::CodeAction", "Pitch Down"]
    }
  }
```

---

> @ovidius72: Organize imports, as far as I know, should run sort imports, remove unused imports and add all missing imports imports` in sequence.

I'm not sure how best to support arbitrary sequences of actions which have be triggered in sequence, but we special cased `editor::OrganizeImports` as a hard coded action here:

- https://github.com/zed-industries/zed/issues/10004
- https://github.com/zed-industries/zed/pull/25793

**Comment by polymorphicengine** [comment]
Hey @notpeter, did you make any progress on this? thanks a lot!

**Comment by notpeter** [comment]
Nope. I no longer work at Zed Industries and never pushed the branch, so someone else will have to start from scratch if they want to pick this up. Sorry.

**Comment by bitfield** [comment]
The lack of this feature is especially time-consuming with extensions like Codebook—there are always lots of false positive spelling errors detected, and I have to go through them like this:

* Go to keyboard and hit F8 for 'next problem' and Cmd-. to invoke 'code actions'
* Go to the mouse to select "Add WORD to dictionary"
* Go back to the keyboard and hit F8...

This is a violation of the prevailing Zed philosophy that everything should be drivable from the keyboard.

**Comment by jdutheil** [comment]
Hey,

I have a working prototype locally for that feature ; basically allowed me to map "CMD + Maj + i" to the "import all missing imports" action like that : 

```
{
    "context": "Editor",
    "bindings": {
      "cmd-shift-i": [
        "editor::PerformCodeAction",
        { "kind": "source.addMissingImports.ts" },
      ],
 }
```

Would be happy to be assigned and work on this feature
